pkg/api/proxy.go

```diff
@@ -0,0 +1,310 @@
+// resulting in a Denial-of-Service (DOS).
+// ref: https://en.wikipedia.org/wiki/SipHash
+func computeTargetMember(config *config.Config, name string) string {
+	h := siphash.New([]byte(config.Cluster.HashKey))
+	h.Write([]byte(name))
+	sum64 := h.Sum64()
+
+	return config.Cluster.Members[sum64%uint64(len(config.Cluster.Members))]
+}
+
+// returns if the target member input is the local instance or not.
+// accepts a list of Local sockets (IP:Port) and a list of target sockets (IP:Port).
+// return true if there is any match, otherwise, false.
+func isTargetMemberLocal(localSockets []string, targetSockets []string) (bool, error) {
+	// if any of the targetMember IPs are present in localIPs, then it is a local member
+	for _, targetSocket := range targetSockets {
+		for _, localSocket := range localSockets {
+			localSplitIndex := strings.LastIndex(localSocket, ":")
+			local := []string{localSocket[:localSplitIndex], localSocket[localSplitIndex+1:]}
+
+			localHost := net.ParseIP(local[0])
+			if localHost == nil {
+				return false, fmt.Errorf("%w: %s", ErrFailedToParseIPAddress, local[0])
+			}
+			localPort := local[1]
+
+			targetSplitIndex := strings.LastIndex(targetSocket, ":")
+			target := []string{targetSocket[:targetSplitIndex], targetSocket[targetSplitIndex+1:]}
+
+			targetHost := net.ParseIP(target[0])
+			if targetHost == nil {
+				return false, fmt.Errorf("%w: %s", ErrFailedToParseIPAddress, target[0])
+			}
+			targetPort := target[1]
+
+			// we need to compare using the net package as IPv6 can be written in multiple forms -
+			// expanded and shortened which are inequal strings.
+			if localHost.Equal(targetHost) && (localPort == targetPort) {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+// gets all the local server sockets (IP:Port).
+func getLocalServerSockets(port string) ([]string, error) {
+	localIPs, err := common.GetLocalIPs()
+	if err != nil {
+		return []string{}, err
+	}
+
+	localSockets := make([]string, len(localIPs))
+	for idx, ip := range localIPs {
+		localSockets[idx] = fmt.Sprintf("%s:%s", ip, port)
+	}
+
+	return localSockets, nil
+}
+
+// gets all the server sockets of a target member (IP:Port).
+// if the input is an IP address, returns the same targetMember in an array.
+// if the input is a host name, performs a lookup and returns the server sockets.
+func getTargetMemberServerSockets(targetMemberSocket string) ([]string, error) {
+	// host is the first part of the target member string
+	targetSplitIndex := strings.LastIndex(targetMemberSocket, ":")
+	target := []string{targetMemberSocket[:targetSplitIndex], targetMemberSocket[targetSplitIndex+1:]}
+	targetHost := target[0]
+	targetPort := target[1]
+
+	addr := net.ParseIP(targetHost)
+	if addr != nil {
+		// this is an IP address, return as is
+		return []string{targetMemberSocket}, nil
+	}
+	// this is a hostname - try to resolve to an IP
+	resolvedAddrs, err := common.GetIPFromHostName(targetHost)
```

Comment:
Looks like this code is not reached by tests. We should cover it.

Reply:
Thanks! I've added code coverage for this.